src/main/scala/mesosphere/marathon/core/task/termination/impl/KillStreamWatcher.scala

```diff
@@ -1,101 +1,112 @@
+    * decommissioned as soon as it is both marked as goal decommissioned, and the instance is considered terminal.
+    *
+    * Were we to only wait for the instance to become terminal, and not for the goal to also be decommissioned, then we
+    * could conclude too early that an instance is decommissioned, as it is possible that some random restart could
+    * precede the receipt of the update goal command.
+    */
+  private[impl] val whenTerminalAndDecommissioned: TerminalPredicate = (instance: Instance, _: Set[Task.Id]) =>
+    instance.state.goal == Goal.Decommissioned && considerTerminal(instance.state.condition)
 
-    val killedViaInstanceChanged =
-      EnrichedSource.eventBusSource(classOf[InstanceChanged], eventStream, bufferSize, overflowStrategy).collect {
-        case event if considerTerminal(event.condition) => event.id
-      }
+  /**
+    * Given instance updates, and a set of instances, emit a set of instanceIds that are not yet expunged or considered
+    * terminal. Excludes instances that are expunged or considered terminal in the initial snapshot.
+    *
+    * @param instanceUpdates InstanceTracker instance state subscription stream
+    * @param instances The instances in question to watch
+    * @param predicate The mechanism to judge instance terminality; are we watching for tasks to be killed, or instance to be decommissioned.
+    * @return
+    */
+  private[impl] def emitPendingTerminal(instanceUpdates: InstanceTracker.InstanceUpdates, instances: Iterable[Instance], predicate: TerminalPredicate): Source[Set[Instance.Id], NotUsed] = {
 
-    val killedViaUnknownInstanceTerminated =
-      EnrichedSource.eventBusSource(
-        classOf[UnknownInstanceTerminated], eventStream, bufferSize, overflowStrategy).map(_.id)
+    val instancesMap: Map[Instance.Id, Set[Task.Id]] =
+      instances.map { i => i.instanceId -> i.tasksMap.values.map(_.taskId).toSet }(collection.breakOut)
 
-    // eagerComplete allows us to toss the right materialized cancellable value
-    val liveInstanceIds = killedViaInstanceChanged.merge(killedViaUnknownInstanceTerminated, eagerComplete = true)
+    instanceUpdates
+      .flatMapConcat {
+        case (snapshot, updates) =>
+          val pendingInstanceIds: Set[Instance.Id] = snapshot.instances.iterator
+            .filter { i => instancesMap.contains(i.instanceId) }
+            .filterNot { i => predicate(i, instancesMap(i.instanceId)) }
+            .map(_.instanceId)
+            .to[Set]
 
-    Source(alreadyConsideredTerminal)
-      .mergeMat(liveInstanceIds)(Keep.right)
+          updates
+            .filter { change => instancesMap.contains(change.id) }
+            .scan(pendingInstanceIds) {
+              case (remaining, deleted: InstanceDeleted) =>
+                remaining - deleted.id
+              case (remaining, InstanceUpdated(instance, _, _)) if predicate(instance, instancesMap.getOrElse(instance.instanceId, Set.empty)) =>
```

Comment:
I think it would be fine to check `instancesMap(instance.instanceId)` since we filter in line 71.